# Practical SQL

This Jupyter notebook will display all SQL queries that I have created as answers to the challenge questions at the end of each chapter of the book "Practical SQL: A Beginner's Guide To Storytelling With Data" by Anthony DeBarros.

The first part of this notebook will require us to import a couple of dependencies as well as personal data that will allow us to link out Jupyter Notebook to PostgreSQL.

In [1]:
import psycopg2
import pandas as pd
from sql_data import db, usr, pwd

In [2]:
# Connecting to postgreSQL database
conn = psycopg2.connect(
    host = "localhost",
    database = db,
    user = usr,
    password = pwd,
    port = 5432
)

def execute_query(connection, query):
    connection.autocommit = True
    cursor = connection.cursor()
    try:
        cursor.execute(query)
        results = cursor.fetchall()
        column_names = [i[0] for i in cursor.description]
        results = pd.DataFrame(results, columns= column_names)
        return results
        print("Query executed succesfully!")
        # Closing the cursor
        cursor.close()
        del cursor
        # Closing the connection
        connection.close()
    except OperationalError as e:
        print(f"The error '{e}' occurred.")      

## Chapter 2: Beginning Data Exploration With Select

Challenge Questions

1. Write a query that lists the schools in alphabetical order along with teachers ordered by last name A-Z

In [3]:
query = """
SELECT school, first_name, last_name
FROM teachers
ORDER BY school, last_name;
"""

execute_query(conn, query)

school first_name last_name
0    F.D. Rossevelt HS        Lee  Reynolds
1    F.D. Rossevelt HS   Kathleen     Roush
2    F.D. Rossevelt HS      Janet     Smith
3  Myers Middle School   Samantha      Bush
4  Myers Middle School     Samuel      Cole
5  Myers Middle School      Betty      Diaz

2. Write a query that finds the one teacher whose first name starts with the letter S and who earns more that $40k


In [4]:
query = """
SELECT *
FROM teachers
WHERE salary > 40000 AND
	first_name ILIKE 's%';
"""

execute_query(conn, query)

id first_name last_name               school   hire_date salary
0   3     Samuel      Cole  Myers Middle School  2005-08-01  43500

3. Rank teachers hired since Jan 1, 2010, ordered by highest paid to lowest

In [5]:
query = """
SELECT *
FROM teachers
WHERE hire_date >= '2010-01-01'
ORDER BY salary DESC;
"""

execute_query(conn, query)

id first_name last_name               school   hire_date salary
0   6   Kathleen     Roush    F.D. Rossevelt HS  2010-10-22  38500
1   1      Janet     Smith    F.D. Rossevelt HS  2011-10-30  36200
2   4   Samantha      Bush  Myers Middle School  2011-10-30  36200

## Chapter 3: Understanding Data Types

Challenge Questions

1. Your company delivers fruit and vegetables to local grocery stores, and you need to track the mileaage driven by each driver each day to a tenth of a mile. Assuming no driver would ever travel more that 999 miles in a day, what would be an appropriate data type for the mileage column in your table. Why?

    Assuming we would only need to track to the nearest tenth of a mile and drivers never drive more than 999 miles, we would want a data type that reflects 4 significant figures. I would choose to use the NUMERIC data type with a precision of 4 and a scale of 1.

2. In the table listing each driver in your company, what are appropriate data types for the drivers' first and last names? Why is it a good idea to separate firs and last names into two columns rather than having one larger name column?

    I would choose the VARCHAR data type for the first and last name columns in my table with an ample amount of characters to suit the two fields - (15) and (30) respectively. This data type is best for minimizing space as it can go up to the indicated character count but if a first or last name does not have that many characters, the program will not pad/fill in the remaining values. It is not adviseable to combine both first and last names into one field for the purpose of saving space - the combined field may need to house extra characters like a space or comma to distinguish the two values - and because a combined field may make future queries harder to perform.

3. Assume you have a text column that includes strings formatted as dates. One of the strings is written as '4//2017'. What will happen when you try to convert that string to the timestamp data type?

    An ERROR message should pop up when trying to convert the text '4//2017' in to a TIMESTAMP as it is lacking the months and is not written in the format required by the TIMESTAMP data type - 'YYYY-MM-DD HH:MM:SS'.

## Chapter 4: Importing and Exporting Data

Challenge Questions

1. Write a WITH statement to include with COPY to handle the import of an imaginary text file whose first couple of rows look like this:

In [6]:
# ---------
# id:movie:actor
# 50:#Mission: Impossible#:Tom Cruise
# ---------

	COPY example_table
	FROM 'C:/RandomDirectory/indicatedtextfile.txt'
	WITH (FORMAT CSV, HEADER, DELIMITER ':', QUOTE '#');


2. Using the table us_counties_2010_ you created and filled in this chapter, export to a CSV file the 20 counties in the United States thathave the most housing units. Make sure you export only each county's name, state, and number of housing units.


	COPY (
		SELECT geo_name, state_us_abbreviation, housing_unit_count
		FROM us_counties_2010
		ORDER BY housing_unit_count DESC
		LIMIT 20
	)
	TO 'C:/RandomDirectory/housing_export.csv'
	WITH (FORMAT CSV, HEADER);

3. Imagine you're importing a file that contains a column with these values:

In [7]:
# ----
# 17519.668
# 20084.461
# 18973.335
# ----

3. Will a column in your target table with data type NUMERIC(3,8) work for these values?

    The data type NUMERIC(3,8) will not work as it has switched the precision and scale values. It should instead be NUMERIC(8,3) to indicate that there should be 8 total digits with only 3 of them being to the right of the decimal.

## Chapter 5: Basic Math And Stats With SQL

Challenge Questions

1. Write a SQL statement for calculating the area of a circle whose radius is 5 inches. Do you need parenthese in your calculation? Why or why not?


In [8]:
query = """
SELECT 3.14 * (5 ^ 2);
"""

execute_query(conn, query)

?column?
0      78.5

    Answer: In this case we do not need parentheses as order of operations will give exponents priority over multiplication. However, parentheses may help make the expression easier to understand.

2. Using the  2010 Census county data, find out which New York state county has the highest percentage of the population that identified as "American Indian/Alaska Native Alone".

In [9]:
query = """
SELECT geo_name AS County,
	state_us_abbreviation,
	p0010001 AS total_population,
	p0010005 AS american_indian_alaska_native_alone,
	(CAST(p0010005 AS NUMERIC(8,1)) / p0010001) * 100 AS Pct_American_Indian
FROM us_counties_2010
WHERE state_us_abbreviation = 'NY'
ORDER BY Pct_American_Indian DESC
LIMIT 1;
"""

execute_query(conn, query)

county state_us_abbreviation  total_population  \
0  Franklin County                    NY             51599   

   american_indian_alaska_native_alone     pct_american_indian  
0                                 3797  7.35866974166166011000

3. Was the 2010 median county population higher in California or New York


In [10]:
query = """
SELECT state_us_abbreviation AS "State",
       percentile_cont(0.5)
	   WITHIN GROUP (ORDER BY p0010001) AS "Median"
FROM us_counties_2010
WHERE state_us_abbreviation IN ('NY', 'CA')
GROUP BY state_us_abbreviation;
"""

execute_query(conn, query)

State    Median
0    CA  179140.5
1    NY   91301.0

     Answer: Based on the output from the query above, California had a higher median county population than New York - 179,140.5 to 91,301 respectively.

## Chapter 6: Joining Tables in a Relational Database

Challenge Questions

1. The table us_counties_2010 contains 3143 rows, and us_counties_2000 has 3141. That reflects the ongoing adjustments to county level geographies that typically result from government decision making. Using appropriate joins and the NULL value, identify which counties don't exist in both tables.


In [11]:
# Answer: The following query reveals that there are 5 counties from Alaska along with Broomfield county are not present in both tables.

query = """
SELECT
	ten.geo_name AS county,
	ten. state_us_abbreviation,
	twok.geo_name
FROM us_counties_2010 AS ten
LEFT JOIN us_counties_2000 AS twok
ON ten.state_fips = twok.state_fips
	AND ten.county_fips = twok.county_fips
WHERE twok.geo_name IS NULL;
"""

execute_query(conn, query)

county state_us_abbreviation geo_name
0          Hoonah-Angoon Census Area                    AK     None
1             Petersburg Census Area                    AK     None
2  Prince of Wales-Hyder Census Area                    AK     None
3               Skagway Municipality                    AK     None
4          Wrangell City and Borough                    AK     None
5                  Broomfield County                    CO     None

2. Using either the median() or precentile_cont() functions, determine the median of the percent change in county population.


In [12]:
query = """
SELECT
	PERCENTILE_CONT(.5) WITHIN GROUP (ORDER BY (ROUND((CAST(ten.p0010001 AS NUMERIC(8,1)) - twok.p0010001) / twok.p0010001 * 100, 1))) AS median_pop_change
FROM us_counties_2010 AS ten
JOIN us_counties_2000 AS twok
ON ten.state_fips = twok.state_fips
	AND ten.county_fips = twok.county_fips;
"""

execute_query(conn, query)

median_pop_change
0                3.2

3. Which county had the greates percentage loss of population between 2000 and 2010?

In [13]:
# Answer: The following query shows that St. Bernard Parish had the greatese population loss.

query = """
SELECT c2010.geo_name AS County,
       c2010.state_us_abbreviation,
       c2010.p0010001 AS pop_2010,
       c2000.p0010001 AS pop_2000,
       c2010.p0010001 - c2000.p0010001 AS raw_change,
       round( (CAST(c2010.p0010001 AS DECIMAL(8,1)) - c2000.p0010001)
           / c2000.p0010001 * 100, 1 ) AS pct_change
FROM us_counties_2010 c2010 INNER JOIN us_counties_2000 c2000
ON c2010.state_fips = c2000.state_fips
   AND c2010.county_fips = c2000.county_fips
ORDER BY pct_change ASC;
"""

execute_query(conn, query)

county state_us_abbreviation  pop_2010  pop_2000  \
0     St. Bernard Parish                    LA     35897     67229   
1         Kalawao County                    HI        90       147   
2       Issaquena County                    MS      1406      2274   
3         Cameron Parish                    LA      6839      9991   
4         Orleans Parish                    LA    343829    484674   
...                  ...                   ...       ...       ...   
3132      Loudoun County                    VA    312311    169599   
3133      Lincoln County                    SD     44828     24131   
3134      Flagler County                    FL     95696     49832   
3135        Pinal County                    AZ    375770    179727   
3136      Kendall County                    IL    114736     54544   

      raw_change pct_change  
0         -31332      -46.6  
1            -57      -38.8  
2           -868      -38.2  
3          -3152      -31.5  
4        -140845      -29.1  
...          ...        ...  
3132      142712       84.1  
3133       20697       85.8  
3134       45864       92.0  
3135      196043      109.1  
3136       60192      110.4  

[3137 rows x 6 columns]

## Chapter 7: Table Design That Works For You

    CREATE TABLE albums (
        album_id bigserial,
        album_catalog_code varchar(100),
        album_title text,
        album_artist text,
        album_time interval,
        album_release_date date,
        album_genre varchar(40),
        album_description text
    );

    CREATE TABLE songs (
        song_id bigserial,
        song_title text,
        song_artist text,
        album_id bigint
    );

Use the tables to answer these questions:

1. Modify these CREATE TABLE statements to include primary and foreign keys plus additional constraints on both tables.

    Answer: 
    
    CREATE TABLE albums (
        album_id bigserial,
        album_catalog_code varchar(80) NOT NULL,
        album_title text NOT NULL,
        album_artist text NOT NULL,
        album_release_date date,
        album_genre varchar(30),
        album_description text,
        CONSTRAINT album_id_key PRIMARY KEY (album_id),
        CONSTRAINT release_date_check CHECK (album_release_date > '1/1/1925')
    );

    CREATE TABLE songs (
        song_id bigserial,
        song_title text NOT NULL,
        song_artist text NOT NULL,
        album_id bigint REFERENCES albums (album_id),
        CONSTRAINT song_id_key PRIMARY KEY (song_id)
);

2. Instead of using column_id as a surrogate key for your primary key, are there any columns in albums that could be useful as a natural key? What would you have to know to decide?

    Answer: 
    
    Album_catalogue_code may be a viable primary key however we would have to know if it is unique across multiple companies or if it is always provided.


3. To speed up queries, which columns are good candidates for indexes?

    Answer: 
    
    Any column that is designated as a primary key should be indexed and so should columns used as foreign keys. In these tables, we should consider indexing the album_id, titles, artists, and album_release_date columns.

## Chapter 8: Extracting Information By Grouping and Summarizing

Challenge Questions

1. What is the pattern in the use of technology in libraries? Both the 2014 and 2009 survey tables contain the columns gpterms and pitusr. Write code to calculate the percent change in the sum of each column over time. Watch out for negative values

In [14]:
# Answer the following query should return the pct change in gpterms and pitusr
query = """
SELECT
    pls14.stabr,
    SUM(pls14.gpterms) AS gpterms_2014,
    SUM(pls09.gpterms) AS gpterms_2009,
    ROUND( (CAST(SUM(pls14.gpterms) AS DECIMAL(10, 1)) - SUM(pls09.gpterms)) /
        SUM(pls09.gpterms) * 100, 2) AS pct_change_gpterms,
    SUM(pls14.pitusr) AS pitusr_2014,
    SUM(pls09.pitusr) AS pitusr_2009,
    ROUND( (CAST(SUM(pls14.pitusr) AS DECIMAL(10, 1)) - SUM(pls09.pitusr)) /
        SUM(pls09.pitusr) * 100, 2) AS pct_change_pitusr
FROM pls_fy2014_pupld14a AS pls14
JOIN pls_fy2009_pupld09a AS pls09
    ON pls14.fscskey = pls09.fscskey
WHERE pls14.gpterms >=0 AND pls09.gpterms >= 0 AND pls14.pitusr >=0 AND pls09.pitusr >= 0
GROUP BY pls14.stabr
ORDER BY pct_change_gpterms DESC, pct_change_pitusr DESC
LIMIT 5
;
"""

execute_query(conn, query)

stabr  gpterms_2014  gpterms_2009 pct_change_gpterms  pitusr_2014  \
0    GU           547            59             827.12        39842   
1    DC          1000           594              68.35      1050623   
2    AK           994           618              60.84       771075   
3    DE           772           487              58.52       622515   
4    ID          1792          1151              55.69      1878131   

   pitusr_2009 pct_change_pitusr  
0        19564            103.65  
1       140251            649.10  
2      1061498            -27.36  
3       451689             37.82  
4      1986141             -5.44

2. Both library survey tables contain a column called obereg. Just as we calculated the percent change in visits grouped by state, do the same to group percent changes in visits by U.S. region using obereg. 

In [15]:
query = """
SELECT
    pls14.obereg,
    SUM(pls14.visits) AS visits_2014,
    SUM(pls09.visits) AS visits_2009,
    ROUND( (CAST(SUM(pls14.visits) AS DECIMAL(10, 1)) - SUM(pls09.visits)) /
        SUM(pls09.visits) * 100, 2) AS pct_change
FROM pls_fy2014_pupld14a AS pls14
JOIN pls_fy2009_pupld09a AS pls09
    ON pls14.fscskey = pls09.fscskey
WHERE pls14.visits >=0 AND pls09.visits >= 0
GROUP BY pls14.obereg
HAVING SUM(pls14.visits) > 50000000
ORDER BY pct_change DESC
LIMIT 5
;
"""

execute_query(conn, query)

obereg  visits_2014  visits_2009 pct_change
0     07     67482370     68703182      -1.78
1     01     87441083     92090682      -5.05
2     04    100458571    110484975      -9.07
3     08    244298987    271264003      -9.94
4     02    232953540    261305024     -10.85

3. Which join type will show you all the rows in both tables, including those without a match? Write such a query and add an IS NULL filter in a WHERE clause to show agencies not included in one or the other table.

    A full outer join will show rows from both tables in a join, regardless of if they are matches. Below is a query that will show the agencies not included in one or the other table:

In [16]:
query = """
SELECT
    pls14.stabr AS state_14,
    pls09.stabr AS state_09
FROM pls_fy2014_pupld14a AS pls14
FULL OUTER JOIN pls_fy2009_pupld09a AS pls09
    ON pls14.fscskey = pls09.fscskey
WHERE pls14.stabr IS NULL OR pls09.stabr IS NULL
ORDER BY pls14.stabr DESC
LIMIT 15
;
"""

execute_query(conn, query)

state_14 state_09
0      None       TX
1      None       NJ
2      None       PR
3      None       TX
4      None       VT
5      None       NC
6      None       NE
7      None       PR
8      None       PR
9      None       NH
10     None       ND
11     None       FL
12     None       PA
13     None       AK
14     None       WI

## Chapter 9: Inspecting and Modifying Data

Challenge Questions

The following questions will use the meat_poultry_egg_inspect table to try and answer the following questions: how many of the plants in the table processs meat, and how many process poultry?

1. Create two new columns called meat_processing and poultry_processing in your meat_poultry_egg_inspect table. Each can be of the type boolean.

    The following queries should create two columns in our meat_poultry_egg_inspect table:

        ALTER TABLE meat_poultry_egg_inspect ADD COLUMN meat_processing BOOLEAN;
        ALTER TABLE meat_poultry_egg_inspect ADD COLUMN poultry_processing BOOLEAN;
2. Using UPDATE, set meat_processing = TRUE on any row where the activities column contains the text "Meat Processing". Do the same update on the poultry processing column but this time look for the the text "Poultry Processing" in activities.

    The following queries should make the updates to our table described above:

In [17]:
query = """
UPDATE meat_poultry_egg_inspect
SET meat_processing = TRUE
WHERE activities ILIKE '%Meat Processing%';

UPDATE meat_poultry_egg_inspect
SET poultry_processing = TRUE
WHERE activities ILIKE '%Poultry Processing%';

SELECT *
FROM meat_poultry_egg_inspect
LIMIT 20;
"""

execute_query(conn, query)

est_number                                            company  \
0                  V3507                 Zollinger Cold Storage Corporation   
1                  M1544                                     J & M Meat Co.   
2                   V971                                     Supervalu Inc.   
3                 M34258                          Beatrice Meat Company LLC   
4                   M675                        Caviness Beef Packers, Ltd.   
5                 M45446                               Center Cut Meats LLC   
6                 M46300  Center for Advancement of Meat Production and ...   
7          M33843+P33843                            Center Road Enterprises   
8                  M8496                               Central Meat Packing   
9                  M8081                            Clem's Custom Cut Meats   
10             M978+V978                            John Volpi and Co., Inc   
11                 M1061                       Happy Valley Processing Inc.   
12  M40268+P40268+V40268                    J & R Natural Meats and Sausage   
13                M45490                          J and L Farm Butcher Shop   
14  M7066A+P7066A+V7066A                                      J Bar B Foods   
15     M7066+P7066+V7066                                      J Bar B Foods   
16           M6636+P6636                              J. C. Peters and Sons   
17                M32184                          Johnson's Premium Sausage   
18           M5536+P5536                                  Banner Creek, LLC   
19     M9542+P9542+V9542                                Lemay and Sons Beef   

                             street          city  st    zip           phone  \
0               699 West 1700 South         Logan  UT  84321  (801) 753-1134   
1                       545 9th St.       Oakland  CA  94607  (510) 839-0380   
2          1400 West Gadsden Street        Quincy  FL  32351  (850) 875-2600   
3                  2595 Grant Blvd.      Beatrice  AL  36425  (251) 575-4021   
4   3255 U.S. Highway 60 PO Box 790      Hereford  TX  79045  (806) 357-2339   
5    11646 Barthel Industrial Drive   Albertville  MN  55301  (763) 497-1660   
6         19087 South Frontage Road          Iowa  LA  70647  (337) 588-5008   
7                   139 Center Road  Eagle Bridge  NY  12057  (518) 677-2702   
8              1120 Kempsville Road    Chesapeake  VA  23320  (757) 547-2161   
9                 181 Virginia Ave.     Lexington  KY  40508  (606) 233-0821   
10             5263 Northrup Avenue      St Louis  MO  63110  (314) 772-8550   
11              206 Hobbs Mill Road       Dearing  GA  30808  (706) 595-3777   
12               3450 Riverside Dr.   Paso Robles  CA  93446  (805) 237-8100   
13                2342 S. Union Rd.        Medway  OH  45341  (987) 475-4584   
14             1078 Highway 90 West        Weimar  TX  78962  (830) 788-7511   
15                  100 Main Street       Waelder  TX  78959  (830) 788-7511   
16            953 Lexington Highway       Loretto  TN  38469  (931) 853-4378   
17            606 South Main Street        Elmira  NY  14904  (607) 733-6877   
18                 619 East 4th St.        Holton  KS  66436  (785) 364-3126   
19          126 Daniel Plummer Road     Goffstown  NH  03045  (603) 622-0022   

    grant_date                                         activities  \
0   1990-02-27  Certification - Export, Identification - Meat,...   
1   1982-05-24                                    Meat Processing   
2   2011-09-16  Certification - Export, Identification - Meat,...   
3   2007-08-23                                    Meat Processing   
4   2005-07-15                    Meat Processing, Meat Slaughter   
5   2015-07-31                                    Meat Processing   
6   2013-07-18                    Meat Processing, Meat Slaughter   
7   2009-11-25                    Meat Processing, Meat Slaughter   
8   1998-08-27                    Meat Processing, M

3. Use the data from the new, update columns to count how many plants perform each type of activity. For a bonus activity, count how many plants perform both activities.

    The following queries should give us the count of plants that process meat, poultry, or both:

In [18]:
query = """
SELECT 
    COUNT(meat_processing) AS count_of_meat_plants,
    COUNT(poultry_processing) AS count_of_poultry_plants
FROM meat_poultry_egg_inspect
"""

execute_query(conn, query)

count_of_meat_plants  count_of_poultry_plants
0                  4764                     3728

In [19]:
query = """
SELECT COUNT(*) as plants_who_process_both
FROM meat_poultry_egg_inspect
WHERE meat_processing = TRUE AND
      poultry_processing = TRUE;
"""

execute_query(conn, query)

plants_who_process_both
0                     3338

## Chapter 10: Statistical Funtions in SQL

1. IN this chapter we saw that the correlation coefficient, or r value, of the variables pct_bachelors_higher and median_hh_income was about 0.68. Write a query using the acs data set to show the correlation between pct_masters_higher and median_hh_income. Is the r value higher or lower? What might explain the difference?

    Answer: The result from our query below shows that the correlation coefficient between the variables pct_masters_higher and median_hh_income is roughly 0.57. This r value is lower than the bachelors_income_r calculated earlier in the chapter. This may be a weaker correlation coefficient because it takes more money/time to earn a masters degree or higher and therefore most households will not have an adult who possesses that level of education.

In [20]:
# The following query should show us the correlation between pct_masters_higher and median_hh_income
query = """
SELECT CORR(median_hh_income, pct_masters_higher) AS masters_income_r
FROM acs_2011_2015_stats;
"""

execute_query(conn, query)

masters_income_r
0            0.5682

2. In the FBI crime data, which cities with a population of 500k or more have the highest rates of motor vehicle thefts (column motor_vehicle_theft)? Which have the highest violent crime rates (column violent_crime)?

    Answer: As shown by our queries below, Milwaukee has the highest rate of motor vehicle theft(12.3%) and Detroit has the highest rate of violent crimes (17.6%).

In [21]:
# Finding highest rate of motor vehicle thefts
query = """
SELECT
    city,
    st,
    population,
    motor_vehicle_theft,
    ROUND((motor_vehicle_theft::NUMERIC / population) * 1000, 1) AS pc_per_1000
FROM fbi_crime_data_2015
WHERE population >= 500000
ORDER BY pc_per_1000 DESC
LIMIT 10;
"""

execute_query(conn, query)


city          st  population  motor_vehicle_theft pc_per_1000
0      Milwaukee   Wisconsin      600400                 7380        12.3
1    Albuquerque  New Mexico      559721                 5179         9.3
2      Baltimore    Maryland      621252                 5526         8.9
3  San Francisco  California      863782                 6915         8.0
4        Detroit    Michigan      673225                 5216         7.7
5       San Jose  California     1031458                 7001         6.8
6         Denver    Colorado      682418                 4365         6.4
7        Houston       Texas     2275221                13560         6.0
8   Indianapolis     Indiana      863675                 4991         5.8
9         Dallas       Texas     1301977                 7594         5.8

In [22]:
# Finding highest rate of violent crime rates
query = """
SELECT
    city,
    st,
    population,
    violent_crime,
    ROUND((violent_crime::NUMERIC / population) * 1000, 1) AS pc_per_1000
FROM fbi_crime_data_2015
WHERE population >= 500000
ORDER BY pc_per_1000 DESC
LIMIT 10;
"""

execute_query(conn, query)

city                    st  population  violent_crime  \
0                 Detroit              Michigan      673225          11846   
1                 Memphis             Tennessee      657936          11449   
2               Milwaukee             Wisconsin      600400           9583   
3               Baltimore              Maryland      621252           9542   
4            Indianapolis               Indiana      863675          11124   
5              Washington  District of Columbia      672228           8084   
6  Nashville Metropolitan             Tennessee      658029           7245   
7            Philadelphia          Pennsylvania     1567810          16132   
8                 Houston                 Texas     2275221          21994   
9             Albuquerque            New Mexico      559721           5406   

  pc_per_1000  
0        17.6  
1        17.4  
2        16.0  
3        15.4  
4        12.9  
5        12.0  
6        11.0  
7        10.3  
8         9.7  
9         9.7

3. Revisit the libraries data in the table pls_fy_2014. Rank library agencies based on the rate of visits per 1000 population (column popu_lsa), and limit the query to agencies serving 250,000 people or more

In [23]:
query = """
SELECT
    libname,
    popu_lsa,
    ROUND(
        (visits::NUMERIC / popu_lsa) * 1000, 1
    ) AS visits_per_1000,
    RANK() OVER(ORDER BY  (visits::NUMERIC / popu_lsa) * 1000 DESC) AS visits_rank
FROM pls_fy2014_pupld14a
WHERE visits >= 0 AND
    popu_lsa >= 250000;
"""

execute_query(conn, query)

libname  popu_lsa visits_per_1000  \
0           CUYAHOGA COUNTY PUBLIC LIBRARY    616527         12962.8   
1                   SEATTLE PUBLIC LIBRARY    640500         10273.9   
2    CENTRAL RAPPAHANNOCK REGIONAL LIBRARY    298007         10147.0   
3             HOWARD COUNTY LIBRARY SYSTEM    293142          9893.5   
4                   ALAMEDA COUNTY LIBRARY    557378          8711.8   
..                                     ...       ...             ...   
192               SANTA ANA PUBLIC LIBRARY    331953           773.4   
193             MONT CO-NORRISTOWN PUB LIB    325368           529.1   
194                  MORRIS COUNTY LIBRARY    492276           442.0   
195          PINAL COUNTY LIBRARY DISTRICT    396237           355.8   
196               ARROWHEAD LIBRARY SYSTEM    314428            15.0   

     visits_rank  
0              1  
1              2  
2              3  
3              4  
4              5  
..           ...  
192          193  
193          194  
194          195  
195          196  
196          197  

[197 rows x 4 columns]